### ISSUE TITLE
boolean value doesn't work properly right after calling reattach() Change Detection;

### ISSUE THREAD

**Issue by N1nJaC0d3** [angular/angular#28444]
# 🐞 bug report

### Affected Package

 The issue is caused by package @angular/....
@angular-devkit/core 7.1.0

### Description
**boolean value doesn't work properly right after calling reattach() Change Detection;Odds** 

I am trying to update boolean value interpolation after calling "**this.cdr.reattach();**"
As you can see with my example code(Stackblitz), the boolean value Do updates properly inside regardless of detached/reattached status and prints it correctly on the Console tab.

**However, the interpolation binding with calling reattach() never works properly only with boolean value when Click the button(```<button (click)="toggleDetach()">```)!
{{isDetached ? 'Detached' : ' Attached'}}**

The boolean value interpolation on the button should have changed on the screen right after reattach() depending on the isDetached value(false/true); 

I shouldn't have put "@Input" on isDetached but I wanted to make sure about Change detection.

I even called "this.cdr.detectChanges();" but Doesn't work! I hope this is not a Bug!

Anyway, I don't know why the boolean value interpolation on the template doesn't work properly after detached from Change Detection mechanism! 

I made an example project on Stackblitz.

Please somebody shows me how to resolve this issue.

This is the example code.

https://stackblitz.com/edit/angular-mh3mqu?file=src%2Fapp%2Fmyzone.component.ts 

```
import { Component, OnInit, Input, ChangeDetectorRef, SimpleChanges } from '@angular/core';

@Component({
	selector: 'my-zone2',
	template: `
    <button (click)="onClick()">Increase</button>
    <button (click)="changeToggle()">ChangeToggle</button>
    <my-comp [value]="myValue" [isDetached]="mytoggle" [canDetach]="mydetach"></my-comp>
  `,
	styles: []
})
export class Zone2Component implements OnInit {
	myValue = 0;
	mytoggle = false;
	mydetach = false;

	constructor() {}

	ngOnInit() {}

	onClick() {
		this.myValue++;
	}

	changeToggle() {
		if (this.mytoggle) {
			console.log('After changeToggle: ', this.mytoggle);
			this.mytoggle = false;
			this.mydetach = false;
			console.log('After changeToggle: ', this.mytoggle);
		} else {
			console.log('Before changeToggle: ', this.mytoggle);
			this.mytoggle = true;
			this.mydetach = true;
			console.log('After changeToggle: ', this.mytoggle);
		}
	}
}

@Component({
	selector: 'my-comp',
	styles: [ '' ],
	template: `
    <div>
      <button (click)="DoNothing()">{{doNothing}}</button>
      <button (click)="toggleDetach()">
      {{isDetached ? 'Detached' : ' Attached'}}
      {{canDetach ? 'canAttach' : ' canDatach'}}
      / {{changedValue}}
      </button>
      <button (click)="justToggle()">
      justToggle:
      {{justToggleValue}} //
      {{isDetached ? 'Detached' : ' Attached'}}
      </button>
      <div>{{value}}</div>
      <div>isDetached? {{isDetached}}</div>
      <input #box (keyup.enter)="onEnter(box.value)">
      <p>{{inputValue}}</p>
    </div>
  `
})
export class My2Component {
	@Input() value: number;
	@Input() isDetached = false;
	doNothing: string = 'noNothing';
	@Input() canDetach = false;
	//canDetach = false;
  changedValue: number = 0;
  justToggleValue: boolean = false;

	constructor(private cdr: ChangeDetectorRef) {}

	DoNothing() {
		console.log('DoNothing!');
		if (this.isDetached) {
			this.doNothing = 'doNothing';
		} else {
			this.doNothing = 'canDoSomething';
		}
	}
	inputValue = '';
	onEnter(typedValue: string) {
		console.log('onEnter-inputValue: ', typedValue);
		this.inputValue = typedValue;
	}

	justToggle() {
		if (this.isDetached) {
      this.justToggleValue = false;
			this.isDetached = false;
		} else {
      this.justToggleValue = true;
			this.isDetached = true;
		}
	}

	toggleDetach() {
		console.log('toggleDetach() !!!');

		if (this.isDetached) {
			console.log(' before ---> ', this.canDetach);
			console.log(' before ---> ', this.isDetached);
			this.isDetached = false;
      this.canDetach = false;
      this.justToggleValue = false;
			++this.changedValue;
			this.cdr.reattach();

			this.cdr.markForCheck();
			this.cdr.detectChanges();

			console.log(' after ---> ', this.isDetached);
			console.log(' after ---> ', this.canDetach);
		} else {
			console.log(' before ---> ', this.canDetach);
			console.log(' before ---> ', this.isDetached);
			this.isDetached = true;
      this.canDetach = true;
      this.justToggleValue = true;
			++this.changedValue;
			this.cdr.detach();
			console.log(' after ---> ', this.isDetached);
			console.log(' after ---> ', this.canDetach);
		}
	}


```

## 🔬 Minimal Reproduction

[https://stackblitz.com/edit/angular-mh3mqu?file=src%2Fapp%2Fmyzone.component.ts ](https://stackblitz.com/edit/angular-mh3mqu?file=src%2Fapp%2Fmyzone.component.ts )

## 🌍  Your Environment

**Angular Version:**
Angular CLI: 7.1.0
Node: 10.13.0
OS: darwin x64
Angular: 
... 

Package                      Version
------------------------------------------------------
@angular-devkit/architect    0.11.0
@angular-devkit/core         7.1.0
@angular-devkit/schematics   7.1.0
@schematics/angular          7.1.0
@schematics/update           0.11.0
rxjs                         6.3.3
typescript                   3.1.6

**Comment by alan-mroczek** [comment]
Hi @N1nJaC0d3. This isn't an issue with Angular itself. It's because of the way your code works.

Conditions are always staying at `falsy` state because you are detaching `ChangeDetector`. You update them to `true`, but right after that you immediately are detaching `ChangeDetector`, so it isn't reached at all. 

This can't be fixed by any means in Angular code, because JavaScript nature is single threaded. You need to leave time for `ChangeDetector ` to catch that change. It can be fixed in your code by simply changing:

`this.cdr.detach()` to `setTimeout(() => this.cdr.detach())`

That way, execution of detaching `ChangeDetector`  is being postponed, and placed on something called `Callback Queue` and will be executed right after synchronous code and all code that was previously on `Callback Queue` will finish . In that time Angular caches your change, and performs view update, and then is detached by your code.

Stackblitz with change that I'm describing: https://stackblitz.com/edit/angular-tcj4df?file=src/app/myzone.component.ts

**Comment by jasonaden** [comment]
Thanks @alan-mroczek for the comments here. I'm closing this as resolved based on the above. If there's further information to make this a bug, please file a new issue.
